code/modules/wiremod/shell/module.dm

```diff
@@ -0,0 +1,146 @@
+		if(!istype(selected_component))
+			return
+		selected_component.signal.set_output(COMPONENT_SIGNAL)
+
+
+/obj/item/circuit_component/mod_adapter_core
+	display_name = "MOD circuit adapter core"
+	desc = "Provides a reference to the MODsuit's occupant and allows the circuit to toggle the MODsuit."
+
+	/// The MODsuit module this circuit is associated with
+	var/obj/item/mod/module/attached_module
+
+	/// The name of the module to select
+	var/datum/port/input/module_to_select
+
+	/// The signal to toggle the suit
+	var/datum/port/input/toggle_suit
+
+	/// The signal to select a module
+	var/datum/port/input/select_module
+
+	/// A reference to the wearer of the MODsuit
+	var/datum/port/output/wearer
+
+	/// The name of the last selected module
+	var/datum/port/output/selected_module
+
+/obj/item/circuit_component/mod_adapter_core/populate_ports()
+	// Input Signals
+	module_to_select = add_input_port("Module to Select", PORT_TYPE_STRING)
+	toggle_suit = add_input_port("Toggle Suit", PORT_TYPE_SIGNAL)
+	select_module = add_input_port("Select Module", PORT_TYPE_SIGNAL)
+	// States
+	wearer = add_output_port("Wearer", PORT_TYPE_ATOM)
+	selected_module = add_output_port("Selected Module", PORT_TYPE_STRING)
+
+/obj/item/circuit_component/mod_adapter_core/register_shell(atom/movable/shell)
+	if(istype(shell, /obj/item/mod/module))
+		attached_module = shell
+	RegisterSignal(attached_module, COMSIG_MOVABLE_MOVED, .proc/on_move)
+
+/obj/item/circuit_component/mod_adapter_core/unregister_shell(atom/movable/shell)
+	UnregisterSignal(attached_module, COMSIG_MOVABLE_MOVED)
+	attached_module = null
+
+/obj/item/circuit_component/mod_adapter_core/input_received(datum/port/input/port)
+	if(!attached_module?.mod)
+		return
+	var/obj/item/mod/module/module
+	for(var/obj/item/mod/module/potential_module as anything in attached_module.mod.modules)
+		if(potential_module.name == module_to_select.value)
+			module = potential_module
+	if(COMPONENT_TRIGGERED_BY(toggle_suit, port))
+		INVOKE_ASYNC(attached_module.mod, /obj/item/mod/control.proc/toggle_activate, attached_module.mod.wearer)
+	if(attached_module.mod.active && module && COMPONENT_TRIGGERED_BY(select_module, port))
+		INVOKE_ASYNC(module, /obj/item/mod/module.proc/on_select)
+
+/obj/item/circuit_component/mod_adapter_core/proc/on_move(atom/movable/source, atom/old_loc, dir, forced)
+	SIGNAL_HANDLER
+	if(istype(source.loc, /obj/item/mod/control))
+		RegisterSignal(source.loc, COMSIG_MOD_MODULE_SELECTED, .proc/on_module_select)
+		RegisterSignal(source.loc, COMSIG_ITEM_EQUIPPED, .proc/equip_check)
+		equip_check()
+	else if(istype(old_loc, /obj/item/mod/control))
+		UnregisterSignal(old_loc, list(COMSIG_MOD_MODULE_SELECTED, COMSIG_ITEM_EQUIPPED))
+		selected_module.set_output(null)
+		wearer.set_output(null)
+
+/obj/item/circuit_component/mod_adapter_core/proc/on_module_select()
+	SIGNAL_HANDLER
+	selected_module.set_output(attached_module.mod.selected_module.name)
```

Comment:
shouldn't you be able to get the selected module from the signal params?

Reply:
Nope. COMSIG_MOD_MODULE_SELECTED has no params.

Reply:
I would have figured that out sooner _if Fikou had documented it_